proposals/p3938.md

```diff
@@ -144,6 +145,35 @@ alias Obj = Other.Obj;
 The `export` keyword is invalid in implementation files, because they do not
 export anything.
 
+### Source file introduction
+
+In the
+[source file introduction](/docs/design/code_and_name_organization/README.md#source-file-introduction),
+the `export` directive is intermingled with `import` directives. It is invalid
+later in the file.
+
+This allows:
+
+```
+import library "foo";
+export FooType;
+
+import library "bar";
+export BarType;
+
+export import library "wiz";
+```
+
+It disallows `export` directives following other code:
+
+```
+import library "foo";
+
+class C { ... };
+
+export FooType;
```

Comment:
I think we should explore the tradeoff here further. We require all `import`s (whether or not they are `export`ed) to be listed first, and I think that's reasonable to preserve, but then it seems we have three choices:

1) As proposed: `export name` can (only) appear in the preamble, with the imports, and cannot appear with the other declarations in the library. This seems potentially a bit unintuitive to me, because other names exported by the library (those the library owns or declares `extern`, or aliases to imported names) can appear later on -- and indeed must.
2) Exact opposite: `export name` can (only) appear after the preamble, so we'd have (a) package declaration, then (b) imports, then (c) declarations that can add new names but not new imports. That seems a bit unsatisfying if you want to group the name exports with the libraries they're exported from.
3) Allow either: `export name` can appear in the preamble, or afterwards.

I *think* the option that seems best to me is (2): treat `export name` like any other non-import declaration. The consideration I find most persuasive here is how we'll handle name lookup within the `name` in `export name` declarations. Either we'll need to carefully process import declarations in the order they appear, intermingled with processing of `export name` declarations, so that the name lookup behavior for `export name` only sees the names that were already imported, *or* name lookup in `export name` syntax will find things that haven't been introduced yet, which seems surprising and inconsistent with our normal behavior for name lookup.

Reply:
I can add this alternative, switching to (2) if that's your preference. To make sure I'm getting this right, trying to note which advantages/disadvantages are particular to each option:

1. `export name` with `import`s

Semantically, names not introduced by the current package are addressed first.

Advantages:

- No need to teach developers they cannot `export` newly introduced names.

Disadvantages:

(mostly the lack of advantages from others)

2. `export name` with other declarations

Semantically, names introduced by by other packages come first, then operations on those names and/or introducing new names.

Advantages:

- `import` remains very special.
- Avoids name lookup questions regarding exporting names added by later `import` directives.

Disadvantages:

- Prevents placing `export name` next to the import that is expected to add the name.
- Means `export import` and `export name` will be in different sections: no single place to look for re-exports.

3. No ordering for `export name`

Semantically, let developers choose what the prefer.

Advantages:

- Maximum flexibility, HOA rule.

Disadvantages:

- Most inconsistent with the desire to treat `import` as special.

Reply:
That sounds good to me. I've asked for feedback on discord, but if there are no opposing opinions by Monday I think we should go ahead with option 2. We can always revisit if we get new information.

Reply:
FYI merged this into the proposal, along with a couple extra thoughts (not sure how you feel about adding a dedicated `export name` section, but I'm expecting a negative response)